File: automations/seo-optimizer/supabase/functions/_shared/orbit.ts
// _shared/orbit.ts
// Wrappers tipados sobre las tablas de Orbit (public.clientes, content_items,
// article_analysis_index) y nuestra seo_optimizer.client_config / v_active_clients.

import { getSupabase, sbSchema } from "./supabase.ts";

export interface OrbitClient {
  id: string;
  name: string;
  gscPropertyUrl: string;
  language: string | null;
  slackChannelId: string | null;
  seoSpecialistUserId: string | null;
  redactorUserId: string | null;
  orbitMainDomain: string | null;
}

export interface ContentItemForAnalysis {
  contentItemId: string;
  clientId: string;
  title: string;
  slug: string | null;
  finalPublishedUrl: string | null;
  language: string | null;
  country: string | null;
  mainKeyword: string;
  secondaryKeywords: string[];
  articleContent: string | null;
  metaDescription: string | null;
  version: number | null;
  isLatestVersion: boolean | null;
  searchIntent: string | null;
  customerJourneyStage: string | null;
  clusterKey: string | null;
  contentRole: string | null;
  mainEntities: unknown;
  semanticFingerprint: string | null;
  primaryKeyword: string | null;
}

export async function listActiveClients(clientIds?: string[] | null): Promise<OrbitClient[]> {
  let q = sbSchema("seo_optimizer").from("v_active_clients").select("*");
  if (clientIds && clientIds.length > 0) q = q.in("client_id", clientIds);
  const { data, error } = await q;
  if (error) {
    console.error("[orbit] listActiveClients failed:", error.message);
    return [];
  }
  return (data ?? []).map(rowToClient);
}

export async function getClient(clientId: string): Promise<OrbitClient | null> {
  const { data, error } = await sbSchema("seo_optimizer")
    .from("v_active_clients").select("*").eq("client_id", clientId).limit(1);
  if (error) {
    console.error("[orbit] getClient failed:", error.message);
    return null;
  }
  if (!data || data.length === 0) return null;
  return rowToClient(data[0]);
}

function rowToClient(r: Record<string, unknown>): OrbitClient {
  return {
    id: r.client_id as string,
    name: (r.client_name as string) ?? "(sin nombre)",
    gscPropertyUrl: (r.gsc_property_url as string) ?? "",
    language: (r.language as string) ?? null,
    slackChannelId: (r.slack_channel_id as string) ?? null,
    seoSpecialistUserId: (r.seo_specialist_user_id as string) ?? null,
    redactorUserId: (r.redactor_user_id as string) ?? null,
    orbitMainDomain: (r.orbit_main_domain as string) ?? null,
  };
}

export async function listPublishedArticlesForClient(clientId: string): Promise<ContentItemForAnalysis[]> {
  const sb = getSupabase();
  const { data: items, error: ciErr } = await sb.from("content_items").select(
    "id, client_id, title, slug, final_published_url, language, country, " +
    "main_keyword, secondary_keywords, article_content, meta_description, " +
    "version, is_latest_version",
  )
    .eq("client_id", clientId)
    .eq("is_latest_version", true)
    .not("final_published_url", "is", null);

  if (ciErr || !items || items.length === 0) {
    if (ciErr) console.warn("[orbit] content_items query failed:", ciErr.message);
    return [];
  }

  const ids = items.map((it: Record<string, unknown>) => it.id as string);
  const { data: idx } = await sb.from("article_analysis_index").select(
    "content_item_id, search_intent, customer_journey_stage, cluster_key, " +
    "content_role, main_entities, semantic_fingerprint, primary_keyword",
  ).in("content_item_id", ids);

  const idxMap = new Map<string, Record<string, unknown>>();
  for (const r of (idx ?? [])) idxMap.set(r.content_item_id as string, r);

  return items.map((it: Record<string, unknown>) => {
    const i = idxMap.get(it.id as string) ?? {};
    return {
      contentItemId: it.id as string,
      clientId: it.client_id as string,
      title: (it.title as string) ?? "",
      slug: (it.slug as string) ?? null,
      finalPublishedUrl: (it.final_published_url as string) ?? null,
      language: (it.language as string) ?? null,
      country: (it.country as string) ?? null,
      mainKeyword: (it.main_keyword as string) ?? "",
      secondaryKeywords: (it.secondary_keywords as string[]) ?? [],
      articleContent: (it.article_content as string) ?? null,
      metaDescription: (it.meta_description as string) ?? null,
      version: (it.version as number) ?? null,
      isLatestVersion: (it.is_latest_version as boolean) ?? null,
      searchIntent: (i.search_intent as string) ?? null,
      customerJourneyStage: (i.customer_journey_stage as string) ?? null,
      clusterKey: (i.cluster_key as string) ?? null,
      contentRole: (i.content_role as string) ?? null,
      mainEntities: i.main_entities ?? null,
      semanticFingerprint: (i.semantic_fingerprint as string) ?? null,
      primaryKeyword: (i.primary_keyword as string) ?? null,
    };
  });
}

export async function getContentItemByUrl(clientId: string, url: string): Promise<ContentItemForAnalysis | null> {
  const sb = getSupabase();
  const { data, error } = await sb.from("content_items").select(
    "id, client_id, title, slug, final_published_url, language, country, " +
    "main_keyword, secondary_keywords, article_content, meta_description, " +
    "version, is_latest_version",
  )
    .eq("client_id", clientId)
    .eq("final_published_url", url)
    .eq("is_latest_version", true)
    .limit(1);

  if (error || !data || data.length === 0) return null;
  const it = data[0];
  return {
    contentItemId: it.id as string,
    clientId: it.client_id as string,
    title: (it.title as string) ?? "",
    slug: (it.slug as string) ?? null,
    finalPublishedUrl: (it.final_published_url as string) ?? null,
    language: (it.language as string) ?? null,
    country: (it.country as string) ?? null,
    mainKeyword: (it.main_keyword as string) ?? "",
    secondaryKeywords: (it.secondary_keywords as string[]) ?? [],
    articleContent: (it.article_content as string) ?? null,
    metaDescription: (it.meta_description as string) ?? null,
    version: (it.version as number) ?? null,
    isLatestVersion: (it.is_latest_version as boolean) ?? null,
    searchIntent: null, customerJourneyStage: null, clusterKey: null,
    contentRole: null, mainEntities: null, semanticFingerprint: null, primaryKeyword: null,
  };
}

File: automations/seo-optimizer/02-agents/_shared/orbit.py
"""Orbit data access — wrappers around public.clientes, public.content_items,
public.article_analysis_index, and seo_optimizer.client_config.

Why this module: Orbit's schema doesn't have everything we need natively:
  - public.clientes columns: id, name, created_at, language, onboarding_manual_tasks
    (NO gsc_property_url, NO status / active flag)
  - We added seo_optimizer.client_config for per-client config + opt-in
  - We read seo_optimizer.v_active_clients which JOINs clientes + client_config
    + proyectos_seo.dominioprincipal (for diagnostics)

If Orbit ever adds these columns natively, simplify by updating this file.
"""
from __future__ import annotations

from dataclasses import dataclass

import structlog

from _shared.supabase_client import sb


log = structlog.get_logger()


@dataclass
class OrbitClient:
    """Active client with all info needed by agents."""
    id: str
    name: str
    gsc_property_url: str
    language: str | None
    slack_channel_id: str | None
    seo_specialist_user_id: str | None
    redactor_user_id: str | None
    orbit_main_domain: str | None


@dataclass
class ContentItemForAnalysis:
    """A content item joined with its analysis index row."""
    content_item_id: str
    client_id: str
    title: str
    slug: str | None
    final_published_url: str | None
    language: str | None
    country: str | None
    main_keyword: str
    secondary_keywords: list[str]
    article_content: str | None
    meta_description: str | None
    version: int | None
    is_latest_version: bool | None
    # From article_analysis_index (LLM-enriched)
    search_intent: str | None
    customer_journey_stage: str | None
    cluster_key: str | None
    content_role: str | None
    main_entities: list | dict | None
    semantic_fingerprint: str | None
    primary_keyword: str | None


# ---------------------------------------------------------------------------
# CLIENTS — via seo_optimizer.v_active_clients
# ---------------------------------------------------------------------------

def list_active_clients(client_ids: list[str] | None = None) -> list[OrbitClient]:
    """Return all clients with seo_optimizer.client_config.is_active = TRUE.

    If client_ids is provided, returns only those (filtered to active ones).
    Reads from view v_active_clients which joins client_config + clientes + proyectos_seo.
    """
    q = sb.schema("seo_optimizer").table("v_active_clients").select("*")
    if client_ids:
        q = q.in_("client_id", client_ids)

    try:
        resp = q.execute()
    except Exception as exc:  # noqa: BLE001
        log.error("v_active_clients_failed", error=str(exc))
        return []

    out: list[OrbitClient] = []
    for r in (resp.data or []):
        out.append(OrbitClient(
            id=r["client_id"],
            name=r.get("client_name") or "(sin nombre)",
            gsc_property_url=r.get("gsc_property_url") or "",
            language=r.get("language"),
            slack_channel_id=r.get("slack_channel_id"),
            seo_specialist_user_id=r.get("seo_specialist_user_id"),
            redactor_user_id=r.get("redactor_user_id"),
            orbit_main_domain=r.get("orbit_main_domain"),
        ))
    return out


def get_client(client_id: str) -> OrbitClient | None:
    """Get one active client by id."""
    try:
        resp = (
            sb.schema("seo_optimizer")
            .table("v_active_clients")
            .select("*")
            .eq("client_id", client_id)
            .limit(1)
            .execute()
        )
    except Exception as exc:  # noqa: BLE001
        log.error("v_active_clients_get_failed", client_id=client_id, error=str(exc))
        return None

    rows = resp.data or []
    if not rows:
        return None
    r = rows[0]
    return OrbitClient(
        id=r["client_id"],
        name=r.get("client_name") or "(sin nombre)",
        gsc_property_url=r.get("gsc_property_url") or "",
        language=r.get("language"),
        slack_channel_id=r.get("slack_channel_id"),
        seo_specialist_user_id=r.get("seo_specialist_user_id"),
        redactor_user_id=r.get("redactor_user_id"),
        orbit_main_domain=r.get("orbit_main_domain"),
    )


# ---------------------------------------------------------------------------
# CONTENT ITEMS + ANALYSIS INDEX
# ---------------------------------------------------------------------------

def list_published_articles_for_client(client_id: str) -> list[ContentItemForAnalysis]:
    """Return all published, latest-version articles for a client, joined with
    their analysis_index row (when present).
    """
    ci_resp = (
        sb.table("content_items")
        .select(
            "id, client_id, title, slug, final_published_url, language, country, "
            "main_keyword, secondary_keywords, article_content, meta_description, "
            "version, is_latest_version"
        )
        .eq("client_id", client_id)
        .eq("is_latest_version", True)
        .not_.is_("final_published_url", "null")
        .execute()
    )
    items = ci_resp.data or []
    if not items:
        return []

    item_ids = [it["id"] for it in items]
    idx_resp = (
        sb.table("article_analysis_index")
        .select(
            "content_item_id, search_intent, customer_journey_stage, cluster_key, "
            "content_role, main_entities, semantic_fingerprint, primary_keyword"
        )
        .in_("content_item_id", item_ids)
        .execute()
    )
    idx_by_id = {r["content_item_id"]: r for r in (idx_resp.data or [])}

    out: list[ContentItemForAnalysis] = []
    for it in items:
        idx = idx_by_id.get(it["id"], {})
        out.append(ContentItemForAnalysis(
            content_item_id=it["id"],
            client_id=it["client_id"],
            title=it.get("title", "") or "",
            slug=it.get("slug"),
            final_published_url=it.get("final_published_url"),
            language=it.get("language"),
            country=it.get("country"),
            main_keyword=it.get("main_keyword", "") or "",
            secondary_keywords=it.get("secondary_keywords") or [],
            article_content=it.get("article_content"),
            meta_description=it.get("meta_description"),
            version=it.get("version"),
            is_latest_version=it.get("is_latest_version"),
            search_intent=idx.get("search_intent"),
            customer_journey_stage=idx.get("customer_journey_stage"),
            cluster_key=idx.get("cluster_key"),
            content_role=idx.get("content_role"),
            main_entities=idx.get("main_entities"),
            semantic_fingerprint=idx.get("semantic_fingerprint"),
            primary_keyword=idx.get("primary_keyword"),
        ))
    return out


def get_content_item_by_url(client_id: str, url: str) -> ContentItemForAnalysis | None:
    """Find a published content_item for a client matching the given URL."""
    resp = (
        sb.table("content_items")
        .select(
            "id, client_id, title, slug, final_published_url, language, country, "
            "main_keyword, secondary_keywords, article_content, meta_description, "
            "version, is_latest_version"
        )
        .eq("client_id", client_id)
        .eq("final_published_url", url)
        .eq("is_latest_version", True)
        .limit(1)
        .execute()
    )
    rows = resp.data or []
    if not rows:
        return None
    it = rows[0]
    return ContentItemForAnalysis(
        content_item_id=it["id"],
        client_id=it["client_id"],
        title=it.get("title", "") or "",
        slug=it.get("slug"),
        final_published_url=it.get("final_published_url"),
        language=it.get("language"),
        country=it.get("country"),
        main_keyword=it.get("main_keyword", "") or "",
        secondary_keywords=it.get("secondary_keywords") or [],
        article_content=it.get("article_content"),
        meta_description=it.get("meta_description"),
        version=it.get("version"),
        is_latest_version=it.get("is_latest_version"),
        search_intent=None,
        customer_journey_stage=None,
        cluster_key=None,
        content_role=None,
        main_entities=None,
        semantic_fingerprint=None,
        primary_keyword=None,
    )
